plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java

```diff
@@ -0,0 +1,647 @@
+import static io.trino.plugin.iceberg.IcebergTableProperties.FILE_FORMAT_PROPERTY;
+import static io.trino.plugin.iceberg.IcebergTableProperties.PARTITIONING_PROPERTY;
+import static io.trino.plugin.iceberg.IcebergUtil.getIcebergTableWithMetadata;
+import static io.trino.plugin.iceberg.IcebergUtil.isHiveMaterializedView;
+import static io.trino.plugin.iceberg.IcebergUtil.loadIcebergTable;
+import static io.trino.plugin.iceberg.PartitionFields.toPartitionFields;
+import static io.trino.spi.StandardErrorCode.ALREADY_EXISTS;
+import static io.trino.spi.StandardErrorCode.INVALID_SCHEMA_PROPERTY;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static io.trino.spi.StandardErrorCode.SCHEMA_NOT_EMPTY;
+import static io.trino.spi.StandardErrorCode.TABLE_NOT_FOUND;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static org.apache.hadoop.hive.metastore.TableType.VIRTUAL_VIEW;
+import static org.apache.iceberg.BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.TABLE_TYPE_PROP;
+import static org.apache.iceberg.TableMetadata.newTableMetadata;
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT_DEFAULT;
+import static org.apache.iceberg.TableProperties.OBJECT_STORE_PATH;
+import static org.apache.iceberg.TableProperties.WRITE_METADATA_LOCATION;
+import static org.apache.iceberg.TableProperties.WRITE_NEW_DATA_LOCATION;
+import static org.apache.iceberg.Transactions.createTableTransaction;
+
+class TrinoHiveCatalog
+        implements TrinoCatalog
+{
+    private static final Logger log = Logger.get(IcebergMetadata.class);
+    private static final String ICEBERG_MATERIALIZED_VIEW_COMMENT = "Presto Materialized View";
+    public static final String DEPENDS_ON_TABLES = "dependsOnTables";
+
+    // Be compatible with views defined by the Hive connector, which can be useful under certain conditions.
+    private static final String TRINO_CREATED_BY = HiveMetadata.TRINO_CREATED_BY;
+    private static final String TRINO_CREATED_BY_VALUE = "Trino Iceberg connector";
+    private static final String PRESTO_VIEW_COMMENT = HiveMetadata.PRESTO_VIEW_COMMENT;
+    private static final String PRESTO_VERSION_NAME = HiveMetadata.PRESTO_VERSION_NAME;
+    private static final String PRESTO_QUERY_ID_NAME = HiveMetadata.PRESTO_QUERY_ID_NAME;
+    private static final String PRESTO_VIEW_EXPANDED_TEXT_MARKER = HiveMetadata.PRESTO_VIEW_EXPANDED_TEXT_MARKER;
+
+    private final CatalogName catalogName;
+    private final HiveMetastore metastore;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final HiveTableOperationsProvider tableOperationsProvider;
+    private final String trinoVersion;
+    private final boolean useUniqueTableLocation;
+
+    private final Map<SchemaTableName, TableMetadata> tableMetadataCache = new ConcurrentHashMap<>();
```

Comment:
This actually brings up the question of scope of `TrinoCatalog`. A Trino `ConnectorMetadata` instance (i.e. `IcebergMetadata`) is created per SQL transaction via `Connector.getMetadata()`. Thus, we store the snapshot isolation info in the `IcebergMetadata` instance.

The snapshot isolation is primarily handled by the `snapshotIds` map in `IcebergMetadata`, to ensure we use the same table snapshots for the entire transaction. I believe that `tableMetadataCache` was added for performance. However, I think since Iceberg (currently?) does not support transactional (historical) metadata, the metadata cache is also required for snapshot isolation.

I would prefer to keep both `snapshotIds` and `tableMetadataCache` in `IcebergMetadata`, since that's the canonical place to store transaction information, and it keeps the catalog implementation simpler.

Going back to the `TrinoCatalog` scope question, what is expected? Do they need to be thread-safe? Could we turn `TrinoCatalogFactory` into a Guice `Provider` and inject `TrinoCatalog` as a singleton?